gobblin-runtime/src/main/java/gobblin/runtime/DistributedJobLauncher.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.runtime;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Queue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Queues;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.util.ExecutorsUtils;
+import gobblin.util.ParallelRunner;
+
+
+/**
+ * An extension to {@link AbstractJobLauncher} for launching a Gobblin job on a distributed
+ * execution framework such as Hadoop MapReduce or Yarn.
+ *
+ * <p>
+ *   This class implements common methods needed to launch and run a Gobblin job on the
+ *   distributed execution framework of choice. For example, one common task is to collect
+ *   output {@link TaskState}s stored in files output by the mappers or containers.
+ * </p>
+ *
+ * @author ynli
+ */
+public abstract class DistributedJobLauncher extends AbstractJobLauncher {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DistributedJobLauncher.class);
+
+
+  protected final FileSystem fs;
+
+  // Number of ParallelRunner threads to be used for state serialization/deserialization
+  protected final int stateSerDeRunnerThreads;
+
+  // Interval in seconds between two runs of the collector of output TaskStates
+  private final int outputTaskStatesCollectorIntervalSeconds;
+
+  // A scheduled executor for the collector of output TaskStates
+  private final ScheduledExecutorService outputTaskStatesCollectorExecutor;
+
+  public DistributedJobLauncher(Properties jobProps, FileSystem fs, Map<String, String> eventMetadata)
```

Comment:
Can you explain what's the benefit of putting this logic in an `AbstractScheduledService` and having to use a `ServiceManager` to properly use it? My thought is this logic is specific to the launching and execution of a job, instead of a generic one that can be used in other places. Additionally we also need to call `collectOutputTaskStates` one last time manually to avoid missing any output `TaskState`s.